packages/eslint-plugin/tests/rules/no-unnecessary-type-parameters.test.ts

```diff
@@ -0,0 +1,485 @@
+    {
+      code: `
+        function third<A, B, C>(a: A, b: B, c: C): C {
+          return c;
+        }
+      `,
+      errors: [
+        { messageId: 'sole', data: { name: 'A' } },
+        { messageId: 'sole', data: { name: 'B' } },
+      ],
+    },
+    {
+      code: `
+        function parseYAML<T>(input: string): T {
+          // ...
+        }
+      `,
+      errors: [{ messageId: 'sole' }],
+    },
+    {
+      code: `
+        function printProperty<T, K extends keyof T>(obj: T, key: K) {
+          console.log(obj[key]);
+        }
+      `,
+      errors: [{ messageId: 'sole', data: { name: 'K' } }],
+    },
+    {
+      code: `
+        class Joiner<T extends string | number> {
+          join(els: T[]) {
+            return els.map(el => '' + el).join(',');
+          }
+        }
+      `,
+      errors: [{ messageId: 'sole' }],
+    },
+    {
+      code: `
+        class Joiner {
+          join<T extends string | number>(els: T[]) {
+            return els.map(el => '' + el).join(',');
+          }
+        }
+      `,
+      errors: [{ messageId: 'sole' }],
+    },
+    {
+      code: `
+        interface Lengthy {
+          length: number;
+        }
+        function getLength<T extends Lengthy>(x: T) {
+          return x.length;
+        }
+      `,
+      errors: [{ messageId: 'sole' }],
+    },
+
+    // These tests are adapted from eslint-plugin-etc
+    {
+      code: `
+        declare function get<T>(): T;
+        get<string>();
+      `,
+      errors: [{ messageId: 'sole' }],
+    },
+    {
+      code: `
+        declare function get<T extends object>(): T;
+      `,
+      errors: [{ messageId: 'sole' }],
+    },
+    {
+      code: `
+        declare function get<T, U = T>(param: U): U;
+      `,
+      errors: [{ messageId: 'sole' }],
+    },
+    {
+      code: `
+        declare function get<T, U extends T = T>(param: T): U;
+      `,
+      errors: [{ messageId: 'sole' }],
+    },
+
+    // These tests are adapted from eslint-plugin-etc's no-misused-generics
+    {
+      code: `
+        // Inferred return type is T, but this is still the sole usage.
+        function fn<T>(param: string) {
+          let v: T = null!;
+          return v;
+        }
+      `,
+      errors: [
+        {
+          data: { name: 'T' },
+          messageId: 'sole',
+        },
+      ],
+    },
+    {
+      code: `
+        declare class C<V> {
+          method<T, U>(param: T): U;
+          prop: <P>() => P;
+        }
+      `,
+      errors: [
+        // V is unused so it's already covered by no-unused-variables
+        { messageId: 'sole', data: { name: 'T' } },
+        { messageId: 'sole', data: { name: 'U' } },
+        { messageId: 'sole', data: { name: 'P' } },
+      ],
+    },
+    {
+      // This is considered valid by eslint-plugin-etc but Dan and Josh disagree;
+      // U only appears once, therefore it's invalid.
```

Comment:
[Aside] Probably good to file an issue on `eslint-plugin-etc` and comment a link here. Maybe they have context we're missing?

If you did and I missed it then it'd be good as a comment here either way.